src/aviary/functional.py

```diff
@@ -0,0 +1,173 @@
+
+        return obs, state_dict
+
+    async def reset(self) -> tuple[Messages, list[Tool]]:
+        obs, extras = await self._call_start()
+        self.state = DynamicState(extras=extras)
+        return [Message(content=obs)], self.tools
+
+    async def step(
+        self, action: ToolRequestMessage
+    ) -> tuple[Messages, float, bool, bool]:
+        msgs = await self.exec_tool_calls(
+            action, state=self.state, concurrency=self.allow_concurrency
+        )
+        return msgs, self.state.reward, self.state.done, False
+
+    def export_frame(self) -> Frame:
+        """Export the current state of the environment."""
+        return Frame(
+            state={
+                "done": self.state.done,
+                "reward": self.state.reward,
+                "extras": self.state.extras,
+            },
+            info={"tool_names": [t.info.name for t in self.tools]},
+        )
+
+
+class EnvironmentBuilder:
+    """Builder class for constructing functional environments."""
+
+    def __init__(self, start_fn: Callable, allow_concurrency: bool):
+        self.start_fn = start_fn
+        self.tools: list[Tool] = []
+        self.allow_concurrency = allow_concurrency
+        update_wrapper(self, start_fn)
+
+    def __call__(self, *args, **kwargs):
+        """Create a new environment instance."""
+        return FunctionalEnvironment(
+            self.start_fn, self.tools.copy(), self.allow_concurrency, *args, **kwargs
+        )
+
+    def tool(self, **tool_kwargs):
+        """Decorator to add tools to the environment."""
+
+        def decorator(func: Callable):
+            tool = Tool.from_function(
+                func, **tool_kwargs, allow_empty_param_descriptions=True
+            )
+            # Check for name conflicts
+            if any(t.info.name == tool.info.name for t in self.tools):
+                raise RuntimeError(f"Tool with name '{tool.info.name}' already exists")
+            self.tools.append(tool)
+            return func
+
+        return decorator
+
+
+class fenv:  # noqa: N801
+    """Factory class for creating functional environments."""
+
+    @staticmethod
+    def start(allow_concurrency: bool = False):
+        """Initialize a new functional environment definition.
+
+        This decorator marks the starting point for defining a functional environment.
+        It should be applied to a function that sets up the initial state and first
+        observation for the environment. The decorated function will serve as the
+        base for adding tools via the .tool() decorator.
+
+        The decorated function should accept any desired parameters that configure
+        the environment instance. It must return a tuple containing:
+        1. The initial observation message (str)
+        2. A dict of any state variables to persist between steps
+
+        Args:
+            allow_concurrency (bool, optional): Whether to allow concurrent tool calls.
```

Comment:
```suggestion
            allow_concurrency (optional): Whether to allow concurrent tool calls.
```

It's best to not duplicate type hints in docstring, as many times here people updated the type hint but forgot about the docstring